### ISSUE TITLE
Provisioned IOPS volumes

### ISSUE THREAD

**Issue by temujin9** [infochimps-labs/ironfan#286]
(Excerpted from https://groups.google.com/a/infochimps.com/forum/?hl=en&fromgroups=#!topic/ironfan-users/2mXPWPrAzCk)

@nickmarden:

> I can imagine a world in which the provider-dependent stuff was inside a cloud (sub)attribute of the volumes directive, or some other similar factorization that allows this functionality while preserving the tidiness of keeping generic volume behaviors separate from provider-specific volume behaviors:
> 
> ```
>     volume :foo do
>       mount_options       'noatime,nodiratime,nobarrier,logbufs=8,logbsize=32k'
>       keep                false
>       device              '/dev/sdm'
>       mount_point         '/data'
>       fstype              'xfs'
>       resizable           true
>       keep                false
>       create_at_launch    true
>       size                50 # GB
>       tags( :mysql_data => true, :persistent => false, :local => false, :bulk => true, :fallback => false )
>       cloud :ec2 do
>         attachable          :ebs
>         snapshot_name       :empty_xfs_small
>         provisioned_iops    ONE_JILLION_PER_SECOND
>       end
>     end
> ```
> 
> I'll let the Infochimps guys tell me how crazy such an idea might be.

@temujin9:

> It's not a horribly crazy idea, but it does have some serious wrinkles to it that would probably require changes to how we interpret the DSL. I'll keep an eye on it when I get to coding for Ironfan v5.

**Comment by nickmarden** [comment]
I was re-examining my proposed syntax and I think there might be an even better way to make this (a) parser-friendly and (b) more compartmentalized:

```
cloud(:ec2) do
  flavor 'm1.xlarge'
  ...
  configuration :super_fast_disk do
    attachable        :ebs
    snapshot_name     :empty_xfs_small
    provisioned_iops  ONE_JILLION_PER_SECOND
  end
  ...
end

volume :foo do
  mount_options       'noatime,nodiratime,nobarrier,logbufs=8,logbsize=32k'
  keep                false
  device              '/dev/sdm'
  mount_point         '/data'
  fstype              'xfs'
  resizable           true
  keep                false
  create_at_launch    true
  size                50 # GB
  tags( :mysql_data => true, :persistent => false, :local => false, :bulk => true, :fallback => false )
  configurations( :super_fast_disk )
end
```

This approach would allow you to define what it means to be a super_fast_disk in a provider-dependent way, but then use that definition in a provider-independent way. It seems that such an approach might be applicable across all sorts of use cases (security groups come to mind!) rather than just volumes.

Just a thought.

**Comment by temujin9** [comment]
I agree: a tight interface between provider and non-provider code does seem a better approach than nesting configurations. Thanks!
